dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java

```diff
@@ -386,6 +381,69 @@ public File downloadTargetEntry(String toolpath, ToolDescriptor.TypeEnum type, b
         }
     }
 
+    /**
+     * Returns a list of all the tool descriptors associated with this workflow
+     * @param type workflow type, CWL, WDL, NFL ...
+     * @param entryPath Workflow path
+     * @param versionID version we are getting descriptors for
+     */
+    public List<ToolFile> getAllToolDescriptors(String type, String entryPath, String versionID) {
+
+        final Ga4Ghv20Api ga4ghv20api = this.client.getGa4Ghv20Api();
+
+        // workflows are identified within ga4gh endpoints when their prefix is '#workflow/' (tools do not require a special prefix)
+        final String ga4ghv20Path = "#workflow/" + entryPath;
+
+        // get all the tool files and filter out anything not a descriptor
+        try {
+            return ga4ghv20api.toolsIdVersionsVersionIdTypeFilesGet(type, ga4ghv20Path, versionID).stream()
+                .filter(toolFile -> toolFile.getFileType().equals(ToolFile.FileTypeEnum.SECONDARY_DESCRIPTOR) || toolFile.getFileType().equals(ToolFile.FileTypeEnum.PRIMARY_DESCRIPTOR))
+                .collect(Collectors.toList());
+        } catch (NullPointerException ex) {
+            exceptionMessage(ex, "Unable to locate entry  " + entryPath + ":" + versionID, Client.COMMAND_ERROR);
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns the version ID of the given workflow, falls back to the latest version
+     * @param entryPath Workflow path
+     */
+    @Override
+    public String getVersionID(String entryPath) {
+        final String[] parts = entryPath.split(":");
+
+        final String versionID = parts.length > 1 ? parts[1] : "master";
+
+        final Workflow workflow = getDockstoreWorkflowByPath(parts[0]);
+
+        // ensure workflow has version
+        Optional<WorkflowVersion> first = workflow.getWorkflowVersions().stream().filter(foo -> foo.getName().equalsIgnoreCase(versionID))
+            .findFirst();
+
+        // if no master is present (for example, for hosted workflows), fail over to the latest descriptor
+        if (first.isEmpty()) {
+            first = workflow.getWorkflowVersions().stream().max(Comparator.comparing(WorkflowVersion::getLastModified));
+            first.ifPresent(workflowVersion -> System.out.println(
+                "Could not locate workflow with version '" + versionID + "'. Using last modified version '" + workflowVersion.getName()
+                    + "' instead."));
+            return first.get().getName();
+        }
+
+        return versionID;
+    }
+
+    /**
+     * Returns if the provided tool path is for a published entry
+     * @param entryPath Tool path
+     */
+    @Override
+    public boolean isEntryPublished(String entryPath) {
+        Workflow workflow = getDockstoreWorkflowByPath(entryPath);
```

Comment:
I don't want to over-optimize this, especially given all the individual checksum calls you'll need to make, and the actual running of the workflow proper, but fetching a workflow is a relatively expensive operation.

Seems like you may be calling it 3 times (I didn't step through it)? If you have to jump through hoops to restructure the code, don't do it, but if it's easy, it would be nicer if you could reuse the Workflow object.

Reply:
I'm happy to refactor. Would it be better to pass the entry object around, or add a variable to the abstract entry client to store the 'current' entry being requested (this could also allow for refactoring elsewhere.)

i.e. if you call `workflow launch --entry entryname --json ...`, I can fetch the workflow specified by `entryname` and store that in the WorkflowClient object. This also means that we could change how other various functions fetch workflows, and potential reduce calls across the CLI. I can also add this as another ticket. Thoughts?

Reply:
This function has been removed. Refactoring further may be challenging, as the [descriptors downloaded](https://github.com/dockstore/dockstore-cli/blob/508cfdb2362f8d01a989b61680a295ac326def5f/dockstore-client/src/main/java/io/dockstore/client/cli/nested/WorkflowClient.java#L353) could potentially be a different version than the one specified in the entryPath. This causes the additional call to [this](https://github.com/dockstore/dockstore-cli/blob/508cfdb2362f8d01a989b61680a295ac326def5f/dockstore-client/src/main/java/io/dockstore/client/cli/nested/AbstractEntryClient.java#L455).